### ISSUE TITLE
How to handle Simulator.close

### ISSUE THREAD

**Issue by drasmuss** [nengo/nengo#931]
Re: discussion in #859

I think we are all in relative agreement that automatically calling `sim.close` (via `__del__` and `atexit`) would be a good thing.

The only debate is, when we do that, would it be good to throw a warning (saying that the user should probably call close manually, rather than letting it be handled automatically).

I think most(?) people would agree that calling close manually is best practices (for the same reason that calling `file.close` is best practices, even though it can be done automatically).  So the question is, do we want to warn users when they are not following best practices?

I think providing a warning in this case is consistent with our general approach in Nengo.  For example, we throw a warning whenever users create a new attribute on a Nengo object, which isn't even really a bad thing to do.  But that warning helps users avoid potentially hard to find bugs, just like the kind you get when an automatic close doesn't happen as you expected.

**Comment by drasmuss** [comment]
(to be clear, I think it is also reasonable to say that we should not give a warning in this case.  I think there are arguments to be made either way, but I lean like 70/30 towards giving a warning)

**Comment by tcstewar** [comment]
> But that warning helps users avoid potentially hard to find bugs, just like the kind you get when an automatic close doesn't happen as you expected.

I think it'd help to have an example in mind of what sort of bug this could be.  I think that may be part of my problem here -- I can't think of any surprising bugs here.  In nengo_spinnaker, it's just that a nice clear Exception gets thrown telling you that there's still a nengo_spinnaker.Simulator around that hasn't been closed yet.  Are there other sorts of bugs people are thinking of?

**Comment by drasmuss** [comment]
Well it's difficult to predict what will happen when a simulator isn't closed properly, across a range of different backends.  Memory leaks, resource conflicts, it will all depend on what that simulator is supposed to be doing when it closes.  Basically if you have a `close` method, we would assume that is because it is important that the simulator be shut down properly.  And if you want the simulator to be shut down properly, something like the context manager syntax is the best way to do that (which is why Python adopted that format for file IO).

**Comment by tbekolay** [comment]
My preference is toward giving a warning too. If this were code I was writing for myself I wouldn't bother, but we've made lots of decisions toward being more verbose about best practices in the past, so it feels like it's in keeping with the rest of Nengo. It's also not really a big ask of our users, as it doesn't add any additional lines of code, and they are already familiar with the `with` syntax from building their model. And like I mentioned before, people who are annoyed can always filter the warning out with `warnings.filter` (I do this for some of the warnings Brian spits out that are irrelevant to me).

> I think it'd help to have an example in mind of what sort of bug this could be. I think that may be part of my problem here.

One example would be the same reason why Python suggests that you close open handles to files and processes -- to be able to free those resources. In my recent learning notebook for ABR, I was doing many parameter sets, and for each parameter set doing multiple sims in parallel. Eventually I got an `IOError: [Errno 24] Too many open files` exception because I had too many handles to open processes; it didn't come up in initial testing, but when I set up a bunch of sims to run overnight, I got the error and had to rerun some things. A similar thing could happen once we add the ability to write to files with probes and don't close the file after running the sim -- again, not a problem for running a simulation or two, but running many sims overnight might fail, which is frustrating.

**Comment by hunse** [comment]
The thing that gets me is what might be fine on one backend might not be fine on another. For example, with the reference backend you can have two simulators open at the same time that you run alternately, no problem. But this might not be possible on SpiNNaker, say. So do we want to be super conservative and always throw errors if more than one simulator is open at the same time? Or throw warnings on the reference backend and errors on SpiNNaker? Or nothing on the reference and errors on SpiNNaker?

**Comment by tcstewar** [comment]
> A similar thing could happen once we add the ability to write to files with probes and don't close the file after running the sim -- again, not a problem for running a simulation or two, but running many sims overnight might fail, which is frustrating.

Nice example...  thanks for this.  And it feels as if it then would interact in interesting ways with the Probe-to-file stuff, and impact our decisions on whether you can still do sim.data[probe] after sim.close() for a probe-to-file (or whether you can do that at all for a probe-to-file...).  However, that makes me think we should hold off on sorting this out until we have those things implemented, so we have actual use cases.

Are there things that already exist in ```nengo_ocl``` that would benefit from a ```sim.close()```?

**Comment by hunse** [comment]
The main benefit of a `sim.close()` in `nengo_ocl` would be that you can free up memory. The same applies to `nengo`, though typically GPU memory is more limited than CPU memory.

**Comment by tcstewar** [comment]
> The main benefit of a sim.close() in nengo_ocl would be that you can free up memory. The same applies to nengo, though typically GPU memory is more limited than CPU memory.

Interesting...  do you think that's a significant amount of memory for standard nengo? I guess it's just the memory for the signals and operators (we can't clean up any of the stuff in sim.data, since we need access to that after sim.close()).  Do we have a sense of how much memory that is?  I'd always just assumed that we're dominated by the encoders and decoders.

**Comment by hunse** [comment]
I think #858 clears up a lot of memory problems, so once that is merged I don't think there will be much benefit in core nengo. I think the way things currently are, we're copying all the encoders and decoders into separate signals, so there could be a significant amount of memory involved.

**Comment by hunse** [comment]
And actually, I think the largest contributor to overall memory usage is eval points. They're huge.

**Comment by tcstewar** [comment]
> And actually, I think the largest contributor to overall memory usage is eval points. They're huge.

Ooo...  good point.  Ouch.  Forgot about those.

I wonder if this sort of thing means that we should have a bit more granularity here....  maybe something like:

```python
sim = Simulator(model)
sim.run(1)
sim.finish()
# now you can access sim.data, but you can't call sim.run()
sim.close()
# now you can't access sim.data, so we can clear up all the memory
```

Although now that I write that, I guess you can get a similar effect like this now:

```python
sim = Simulator(model)
sim.run(1)
sim.close()
# now you can access sim.data, but you can't call sim.run()
del sim
# now you can't access sim.data, so we can clear up all the memory
```

**Comment by hunse** [comment]
Yeah, I like the latter model better.

If we're really worried about the eval points specifically, we could have some option not to store them, since they're not actually needed once the decoders are found. The only reason we keep them around is in case users want to see them.

**Comment by tbekolay** [comment]
I made PR #932 for some implementation if we want to discuss that. Still a WIP, mainly made it to have a place to try out implementation stuff as we discuss in this issue.

**Comment by tcstewar** [comment]
> If we're really worried about the eval points specifically, we could have some option not to store them, since they're not actually needed once the decoders are found. The only reason we keep them around is in case users want to see them.

We could always throw them out, but then re-generate them if the user happens to ask for them...  all we'd need to keep is the seed....   

(I'm not quite sure whether I'm serious about this suggestion...  It's kinda cute, but might be too much black magic....  still, it'd be a pretty big memory savings...)

**Comment by hunse** [comment]
It's a bit more complicated than keeping a seed, since we actually just have the RNG, so we'd need to freeze the whole RNG state (basically copy the RNG and store the copy). It's an interesting idea, but I also worry about it getting too magicky or just downright complicated.

**Comment by jgosmann** [comment]
As `Simulator.close` has been implemented, it seems that this discussion in concluded.

**Comment by hunse** [comment]
I'm not sure if everything discussed here has been concluded...

**Comment by jgosmann** [comment]
If there are still open questions, it seems more useful to me to open a new issue for that. I couldn't find anything by skimming the thread. Except for #933 which is already it's own issue.
